examples/xds-management-server/pkg/providers/file.go

```diff
@@ -0,0 +1,309 @@
+package providers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	envoylistener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	gogojsonpb "github.com/gogo/protobuf/jsonpb"
+	prototypes "github.com/gogo/protobuf/types"
+	"google.golang.org/protobuf/encoding/protojson"
+	"quilkin.dev/xds-management-server/pkg/cluster"
+	"quilkin.dev/xds-management-server/pkg/filterchain"
+	"sigs.k8s.io/yaml"
+
+	"github.com/cenkalti/backoff"
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+)
+
+// FilterConfig represents a filter's config.
+type FilterConfig interface{}
+type resources struct {
+	Clusters    []cluster.Cluster
+	FilterChain []FilterConfig
+}
+
+// FileProvider watches a file on disk for resources.
+type FileProvider struct {
+	configFilePath string
+	proxyIDCh      <-chan string
+}
+
+// NewFileProvider creates a new FileProvider.
+func NewFileProvider(configFilePath string, proxyIDCh chan string) *FileProvider {
+	return &FileProvider{
+		configFilePath: configFilePath,
+		proxyIDCh:      proxyIDCh,
+	}
+}
+
+// Run runs the FileProvider.
+func (p *FileProvider) Run(ctx context.Context, logger *log.Logger) (
+	<-chan []cluster.Cluster, <-chan filterchain.ProxyFilterChain, <-chan error) {
+	clusterCh := make(chan []cluster.Cluster, 10)
+	filterChainCh := make(chan filterchain.ProxyFilterChain, 10)
+	resourcesCh := make(chan resources)
+
+	errorCh := make(chan error)
+
+	go runUpdater(
+		ctx,
+		logger,
+		p.proxyIDCh,
+		resourcesCh,
+		clusterCh,
+		filterChainCh)
+
+	go runConfigFileProvider(
+		ctx,
+		logger,
+		p.configFilePath,
+		resourcesCh,
+		errorCh)
+
+	return clusterCh, filterChainCh, errorCh
+}
+
+func runUpdater(
+	ctx context.Context,
+	logger *log.Logger,
+	proxyIDCh <-chan string,
+	resourcesCh <-chan resources,
+	clusterCh chan<- []cluster.Cluster,
+	filterChainCh chan<- filterchain.ProxyFilterChain,
+) {
+	defer func() {
+		close(clusterCh)
+		close(filterChainCh)
+	}()
+	ticker := time.NewTicker(100 * time.Millisecond)
```

Comment:
Non blocking: 🤔 each time I see this I want to reach for a time mock like github.com/jonboulle/clockwork. (mention this in snapshot.go as well). You're processing relatively fast, so controlling time is likely not an issue during unit tests, but something to consider.

(Possibly another reason to make `runUpdater` a method on `FileProvider`?)